keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,510 @@
+- **Connectivity service** A component built into the KAS which provides a golang dialer for outgoing connection requests.
+The dialer provided depends on NetworkContext information.
+- **Connectivity proxy** The proxy that runs in the master network.
+It has a secure channel established to the cluster network.
+It exposes a gRPC interface to KAS to provide connectivity service.
+- **Flat Network** A network where traffic can be successfully routed using IP.
+Implies no overlapping (i.e. shared) IPs on the network.
+
+## Proposal
+
+We will run a connectivity proxy inside the master network.
+The connectivity proxy exposes a gRPC interface to the KAS.
+Each connectivity proxy allows secure connections to one or more cluster networks.
+Any network addressed by a connectivity proxy must be flat.
+Currently the only mechanism for handling overlapping IP ranges in Kubernetes is the Proxy.
+Non IP routable traffic, past the proxy, would need to be a non Kubernetes mechanism to route.
+
+Running the connectivity proxy in a separate process has a few advantages.
+- The connectivity proxy can be extended without recompiling the KAS.
+Administrators can run their own variants of the connectivity proxy.
+- Traffic can be audited or forwarded (eg. via a proprietary VPN) using a custom connectivity proxy.
+- The separation removes master <-> cluster connectivity concerns from the KAS.
+- The code and responsibility separation lowers the complexity of the KAS code base.
+- The separation reduces the effects of issue such as crashes in the connectivity impacting the KAS.
+Connectivity issues will not stop the KAS from serving API requests.
+This is important as serving API requests may be necessary in order to fix the crashes.
+A problem with a node, set of nodes or load-balancers configuration, may be fixed with API requests.
+
+![API Server Network Proxy Simple Cluster](NetworkProxySimpleCluster.png)
+The diagram shows API Server’s outgoing traffic flow.
+The user (in blue box), master network (in purple cloud) and
+a cluster network (in green cloud) are represented.
+
+The user (blue) initiates communication to the KAS.
+The KAS then initiates connections to other components.
+It could be node/pod/service in cluster networks (red dotted arrow to green cloud),
+or etcd for storage in the same master network (blue arrow) or mutate the request
+based on an admission web-hook (red dotted arrow to purple cloud).
+The KAS handles these cases based on NetworkContext based traffic routing.
+The connectivity proxy should be able to do routing solely based on IP.
+The proxy should not require the NetworkContext. This means the service CIDR,
+node CIDR and pod CIDR of each cluster network cannot overlap.
+
+### Network Context
+
+The minimal NetworkContext looks like the following struct in golang:
+
+```go
+type NetworkContext struct {
+        // NetworkName is the unique name of the network
```

Comment:
It's a little confusing that this field is NetworkName, but the field that it matches below is connectionService.name

Either the below should have NetworkName to match or we need to do some better alignment. Otherwise, it will be hard for users to discover the relationship.

Reply:
Fixed